client/lib/importer/store.js

```diff
@@ -46,168 +43,233 @@ const initialState = Immutable.fromJS( {
+					...state.api,
+					isFetching: false,
+					isHydrated: true,
+					retryCount: 0,
+				},
+			};
 
 		case IMPORTS_IMPORT_CANCEL:
 		case IMPORTS_IMPORT_RESET:
-			// Remove the specified importer from the list of current importers
-			newState = state.update( 'importers', importers => {
-				return importers.filterNot(
-					importer => importer.get( 'importerId' ) === action.importerId
-				);
-			} );
-			break;
-
-		case IMPORTS_UPLOAD_FAILED:
-			newState = state
-				.setIn( [ 'importers', action.importerId, 'importerState' ], appStates.UPLOAD_FAILURE )
-				.setIn( [ 'importers', action.importerId, 'errorData' ], {
-					type: 'uploadError',
-					description: action.error,
-				} );
-			break;
-
+			return {
+				...state,
+				importers: {
+					...omit( state.importers, action.importerId ),
+				},
+			};
+
+		case IMPORTS_UPLOAD_FAILED: {
+			const { importerId } = action;
+			const importerItem = getImporterItemById( state, importerId );
+
+			return {
+				...state,
+				importers: {
+					...state.importers,
+					[ importerId ]: {
+						...importerItem,
+						importerState: appStates.UPLOAD_FAILURE,
+						errorData: {
+							type: 'uploadError',
+							description: action.error,
+						},
+					},
+				},
+			};
+		}
 		case IMPORTS_UPLOAD_COMPLETED:
-			newState = state
-				.deleteIn( [ 'importers' ], action.importerId )
-				.setIn(
-					[ 'importers', action.importerStatus.importerId ],
-					Immutable.fromJS( action.importerStatus )
-				);
-			break;
-
-		case IMPORTS_AUTHORS_START_MAPPING:
-			newState = state.setIn(
-				[ 'importers', action.importerId, 'importerState' ],
-				appStates.MAP_AUTHORS
-			);
-			break;
-
-		case IMPORTS_AUTHORS_SET_MAPPING:
-			newState = state.updateIn(
-				[ 'importers', action.importerId, 'customData', 'sourceAuthors' ],
-				authors =>
-					authors.map( author => {
-						if ( action.sourceAuthor.id !== author.get( 'id' ) ) {
-							return author;
-						}
-
-						return author.set( 'mappedTo', action.targetAuthor );
-					} )
-			);
-			break;
-
-		case IMPORTS_IMPORT_RECEIVE:
-			newState = state.setIn( [ 'api', 'isHydrated' ], true );
-
-			if ( newState.getIn( [ 'importerLocks', action.importerStatus.importerId ], false ) ) {
-				break;
-			}
-
-			if ( action.importerStatus.importerState === appStates.DEFUNCT ) {
-				newState = newState.deleteIn( [ 'importers', action.importerStatus.importerId ] );
-				break;
+			return {
+				...state,
+				importers: {
+					...omit( state.importers, action.importerId ),
+					[ action.importerStatus.importerId ]: action.importerStatus,
+				},
+			};
+
+		case IMPORTS_AUTHORS_START_MAPPING: {
+			const { importerId } = action;
+			const importerItem = getImporterItemById( state, importerId );
+
+			return {
+				...state,
+				importers: {
+					...state.importers,
+					[ importerId ]: {
+						...importerItem,
+						importerState: appStates.MAP_AUTHORS,
+					},
+				},
+			};
+		}
+		case IMPORTS_AUTHORS_SET_MAPPING: {
```

Comment:
Would be nice to regularise the line breaks between cases in this file.

Reply:
hey are somewhat regular in that those with lexical scope blocks close before the next case, those without don't have the closing bracket to do the same so a gap is left in place.
Perhaps we could leave a gap between the closing block and the next case?

```js
}

case IMPORTS_AUTHORS_SET_MAPPING: {
```

^ Would be a little more uniform :)

I'd actually be much in favour of a linting rule that enforces lexical scope on all switch cases - we have a number of cases (no pun intended) where variables are defined inside of a case without proper scoping.

Reply:
Oh I see, sorry!